### DISCUSSION TITLE
Implement cURL for many small files and compatibility

### DISCUSSION THREAD

**Discussion by silveroxides** [BlafKing/sd-civitai-browser-plus#261] [Ideas]
aria2 might be good for downloading larger files or larger archives of multiple files.
But it is horribly slow when it comes to downloading lots of images for previews and smaller requests.
[Link to feature comparison table](https://curl.se/docs/comparison-table.html)
cURL is also present in every OS by default and does not require additional binaries to run.
I have been holding back on even updating any of my model tags, metadata and previews just cause of how slow it is.
Just finished images. took 15 minutes. 250/250Mbit/s

**Comment by BlafKing** [comment]
Sounds good, I'll have a look at implementing it.

It's worth noting that Aria2 is only used for downloading models, not for updating tags, metadata or previews.
Those are handled by the `requests` library from python which uses `urllib3`

**Comment by BlafKing** [comment]
After looking around, there isn't a huge speed improvement to be gained from just switching to `cURL` instead of `urllib3`.

The performance of HTTP requests, whether made with `cURL` or `urllib3`, is often limited by the time it takes for the server to respond rather than the client-side request handling.

The wait time you're experiencing mostly comes from waiting for a response from the CivitAI API.
Increasing the speed for preview image downloads would require me to introduce asynchronous downloads for the images, I could look into this in the future, but it'll likely be tricky to implement as the current codebase wasn't built with asynchronicity in mind.

Besides, downloading a lot of preview images is usually something you'll have to do only once for all the models that were not downloaded with this extension.

Last note, the speed for updating just the tags and metadata should already be really quick, for me updating 240 tags and metadata took only a few seconds
(it takes longer if you're also generating HTML files, which is caused by having to do an extra API request for each HTML file.)

Considering all of this, I won't be implementing `cURL` over the current `requests` library.
Thanks for the suggestion though!
